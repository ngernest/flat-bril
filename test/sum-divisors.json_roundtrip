{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "i",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "res",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "non",
          "op": "const",
          "type": "int",
          "value": -1
        },
        {
          "dest": "zer",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "n",
            "zer"
          ],
          "dest": "nlz",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "nlz"
          ],
          "funcs": [],
          "labels": [
            "neg",
            "begin"
          ],
          "op": "br"
        },
        {
          "label": "neg"
        },
        {
          "args": [
            "n",
            "non"
          ],
          "dest": "n",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "label": "begin"
        },
        {
          "args": [
            "i",
            "one"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "i",
            "i"
          ],
          "dest": "isq",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "isq",
            "n"
          ],
          "dest": "sqgt",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "sqgt"
          ],
          "funcs": [],
          "labels": [
            "end",
            "check"
          ],
          "op": "br"
        },
        {
          "label": "check"
        },
        {
          "args": [
            "n",
            "i"
          ],
          "dest": "m",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "m",
            "zer"
          ],
          "dest": "eqz",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "eqz"
          ],
          "funcs": [],
          "labels": [
            "body",
            "begin"
          ],
          "op": "br"
        },
        {
          "label": "body"
        },
        {
          "args": [
            "i"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "res",
            "i"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "n",
            "i"
          ],
          "dest": "d",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "d",
            "i"
          ],
          "dest": "deqi",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "deqi"
          ],
          "funcs": [],
          "labels": [
            "begin",
            "then"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "args": [
            "d"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "res",
            "d"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "begin"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "res"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "dividend",
          "type": "int"
        },
        {
          "name": "divisor",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "dividend",
            "divisor"
          ],
          "dest": "quot",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "divisor",
            "quot"
          ],
          "dest": "prod",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "dividend",
            "prod"
          ],
          "dest": "diff",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "diff"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
