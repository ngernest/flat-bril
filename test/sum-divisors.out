@main
	op: const		dest: i 		type: int		value: 0	
	op: const		dest: res		type: int		value: 0	
	op: const		dest: one		type: int		value: 1	
	op: const		dest: non		type: int		value: -1	
	op: const		dest: zer		type: int		value: 0	
	op: lt   		dest: nlz		type: bool		args: ["n", "zer"]	
	op: br   		args: ["nlz"]		labels: ["neg", "begin"]
.neg
	op: mul  		dest: n 		type: int		args: ["n", "non"]	
.begin
	op: add  		dest: i 		type: int		args: ["i", "one"]	
	op: mul  		dest: isq		type: int		args: ["i", "i"]	
	op: gt   		dest: sqgt		type: bool		args: ["isq", "n"]	
	op: br   		args: ["sqgt"]		labels: ["end", "check"]
.check
	op: call 		dest: m 		type: int		args: ["n", "i"]		funcs: ["mod"]
	op: eq   		dest: eqz		type: bool		args: ["m", "zer"]	
	op: br   		args: ["eqz"]		labels: ["body", "begin"]
.body
	op: print		args: ["i"]	
	op: add  		dest: res		type: int		args: ["res", "i"]	
	op: div  		dest: d 		type: int		args: ["n", "i"]	
	op: eq   		dest: deqi		type: bool		args: ["d", "i"]	
	op: br   		args: ["deqi"]		labels: ["begin", "then"]
.then
	op: print		args: ["d"]	
	op: add  		dest: res		type: int		args: ["res", "d"]	
	op: jmp  		labels: ["begin"]
.end
	op: print		args: ["res"]	
@mod
	op: div  		dest: quot		type: int		args: ["dividend", "divisor"]	
	op: mul  		dest: prod		type: int		args: ["divisor", "quot"]	
	op: sub  		dest: diff		type: int		args: ["dividend", "prod"]	
	op: ret  		args: ["diff"]	
